source/common/grpc/async_client_impl.cc

```diff
@@ -251,7 +260,9 @@ void AsyncStreamImpl::onData(Buffer::Instance& data, bool end_stream) {
 void AsyncStreamImpl::onTrailers(Http::ResponseTrailerMapPtr&& trailers) {
   auto grpc_status = Common::getGrpcStatus(*trailers);
   const std::string grpc_message = Common::getGrpcMessage(*trailers);
-  callbacks_.onReceiveTrailingMetadata(std::move(trailers));
+  if (!waiting_to_delete_on_remote_close_) {
+    callbacks_.onReceiveTrailingMetadata(std::move(trailers));
```

Comment:
Similarly, I think we should surface this trailers to upper layer, such that app-layer get a chance to look at the overall grpc-status (and do their stuff).